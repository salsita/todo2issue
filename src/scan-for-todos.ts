import { readFile } from 'fs/promises'
import { extname, resolve } from 'path'
import { parse } from 'leasot'
import { Todo } from './model'

const ISSUE_NUMBER_PATTERN = /^#([0-9]+)$/

export async function scanForTodos(root: string, files: string[]): Promise<Todo[]> {
  const todos: Todo[] = []
  for (const filename of files) {
    const contents = await readFile(resolve(root, filename), 'utf8')
    const todoComments = parse(contents, { filename, extension: extname(filename) })

    todos.push(
      ...todoComments.map((todoComment) => {
        const { line, text, tag, ref } = todoComment
        const [, issueNumber] = ISSUE_NUMBER_PATTERN.exec(ref) || []
        const todo = {
          filename,
          line,
          tag,
          text,
          issueNumber: issueNumber ? Number(issueNumber) : undefined,
        }
        return todo
      }),
    )
  }
  return todos
}
