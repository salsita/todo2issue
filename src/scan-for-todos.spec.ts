import { scanForTodos } from './scan-for-todos'
import { resolve } from 'path'

describe('scanForTodos', () => {
  it('finds all TODOs', async () => {
    const todos = await scanForTodos(resolve(__dirname, '../sample'), ['src/index.ts'])
    console.log(JSON.stringify(todos, null, 2))
    expect(todos).toMatchSnapshot()
  })
})
