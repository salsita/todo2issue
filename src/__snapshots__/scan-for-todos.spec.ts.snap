// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanForTodos finds all TODOs 1`] = `
Array [
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 1,
    "tag": "TODO",
    "text": "Single line todo comment",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": 123,
    "line": 2,
    "tag": "TODO",
    "text": "Single line todo comment with reference",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": 123,
    "line": 4,
    "tag": "TODO",
    "text": "Same reference different comment",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 5,
    "tag": "TODO",
    "text": "Same reference different comment",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 7,
    "tag": "TODO",
    "text": "Multiline todo",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": 124,
    "line": 12,
    "tag": "TODO",
    "text": "Multiline todo",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 15,
    "tag": "TODO",
    "text": "Multiple single",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 16,
    "tag": "TODO",
    "text": "line comments",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": 125,
    "line": 19,
    "tag": "TODO",
    "text": "Multiple single",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 20,
    "tag": "TODO",
    "text": "line comments with reference",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 22,
    "tag": "TODO",
    "text": "With colon",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 24,
    "tag": "TODO",
    "text": "With colon",
  },
  Object {
    "filename": "src/index.ts",
    "issueNumber": undefined,
    "line": 26,
    "tag": "TODO",
    "text": "With colon",
  },
]
`;
